transaction.go

```diff
@@ -21,42 +27,118 @@ type TransactionsRequest struct {
-	Offset            int
-	Limit             int
+	Client            *Client
+}
+
+type TransactionsIterator struct {
+	request   *TransactionsRequest
+	response  *TransactionsResponse
+	batchSize int
 }
 
 type TransactionsResponse struct {
 	Errors  []map[string]string `json:"errors"`
 	Results []Transaction       `json:"results"`
-	Pages   map[string]string   `json:"pages"`
+	Pages   Pages               `json:"pages"`
+}
+
+func (t *TransactionsRequest) GetAll() ([]Transaction, error) {
+	var resp *TransactionsResponse
+	var err error
+	if resp, err = t.get(nil); err != nil {
+		return []Transaction{}, err
+	}
+
+	return resp.Results, nil
 }
 
-func (c *Client) GetTransactions(transactionsReq TransactionsRequest) ([]Transaction, error) {
+func (t *TransactionsRequest) get(paginationParams *url.Values) (*TransactionsResponse, error) {
 	var err error
 	var req *http.Request
-	var transactions []Transaction
+	var response *TransactionsResponse
+
+	params := &url.Values{}
+	if t.CheckingAccountID != "" {
+		params.Set("checking_account_id", t.CheckingAccountID)
+	}
+	if t.Status != "" {
+		params.Set("status", t.Status)
+	}
+	dateLayout := "2006-01-02"
+	if !t.From.IsZero() {
+		params.Set("from", t.From.Format(dateLayout))
+	}
+	if !t.To.IsZero() {
+		params.Set("to", t.To.Format(dateLayout))
+	}
+
+	var url *url.URL
+	if url, err = url.Parse(fmt.Sprintf("%s/%s", ApiBase, "transactions/")); err != nil {
+		return nil, err
+	}
+
+	if paginationParams != nil {
+		url.RawQuery = fmt.Sprintf("%s&%s", params.Encode(), paginationParams.Encode())
+	}
 
-	if req, err = http.NewRequest("GET", fmt.Sprintf("%s/%s", ApiBase, "transactions/"), nil); err != nil {
-		return transactions, err
+	if req, err = http.NewRequest("GET", url.String(), nil); err != nil {
+		return nil, err
 	}
 	var resp *http.Response
 
-	if resp, err = c.Do(req); err != nil {
-		return transactions, err
+	if resp, err = t.Client.do(req); err != nil {
+		return nil, err
 	}
 
-	var response TransactionsResponse
-	if err = json.NewDecoder(resp.Body).Decode(&response); err != nil {
-		return transactions, err
+	if err = json.NewDecoder(resp.Body).Decode(response); err != nil {
+		return nil, err
 	}
+	return response, nil
+}
 
-	return transactions, nil
+func (r *TransactionsRequest) Iterator() TransactionsIterator {
+	return TransactionsIterator{
+		request:   r,
+		batchSize: MaxBatchSize,
+	}
 }
 
-func (c *Client) Do(req *http.Request) (*http.Response, error) {
-	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", c.accessToken))
-	if c.clientVersion != "" {
-		req.Header.Set("Client-Version-Id", c.clientVersion)
+func (t *TransactionsIterator) SetBatchSize(n int) {
+	if n < MaxBatchSize {
+		t.batchSize = n
+	}
+}
+
+func (t *TransactionsIterator) Next() ([]Transaction, []map[string]string, error) {
+	var err error
+	var params *url.Values
+	if t.response != nil {
+		params = t.response.Pages.Next
+	} else {
+		params = &url.Values{"limit": []string{strconv.Itoa(t.batchSize)}}
```

Comment:
Just to double check, does this handle correctly when the user doesn't specify a batch size? as in, will it still work, or would we just fetch 0 transactions since limit is `0`?

I'm thinking a sane fallback would be to just fetch everything, or a fixed constant, as long as it's not fetching nothing